### DISCUSSION TITLE
python linting and members

### DISCUSSION THREAD

**Discussion by crankywilson** [microsoft/pylance-release#2206] [Q&A] [answered]
Hi, I don't know a whole lot about python linting, but every tool I've tried has failed the following case:

    class PyClass:
      def __init__(self):
        self.member1 = 1
        self.member2 = 2
    
    def someFunction():
      inst = PyClass()
      print(inst.nonExistentMember)
    
    someFunction()

I want the linter to tell me when references to non-existent members are found.  This comes up when making changes to classes after code already references the pre-changed class, and I'm used to compilers being able to tell me where I need to update in other languages. 

Pylance has no problem with the above code, nor does mypy or other tools I've tried.  In fact Pylance is ok with moving the code out of someFunction although mypy complains if the bad reference is in the main script scope.  I assume the problem is because the linters are not branching into functions or something?

But the whole point of linting, to me, or one of the points, is to assume that python classes are going to be assumed 'static' like in traditional typed languages - once I define the members, any non-defined member encountered should be considered an error?

Is there a way to get the behavior I'm looking for?

**Comment by erictraut** [comment, marked as answer]
The term "linting" typically refers to tools that check for code style conformance (indentation, naming, etc.). By that definition, Pylance doesn't do any linting. For that, you can enable pylint or other tools specifically focused on code style.

What you're looking for is static type checking, and Pylance does provide that. However, type checking errors are disabled by default in Pylance because many users don't want to be bothered by such errors. To enable this, set the "python.analysis.typeCheckingMode" setting to "basic". Once this mode is enabled, Pylance will emit errors or warnings for things like accesses to attributes that are not declared within a class.

<img width="658" alt="Screen Shot 2022-01-02 at 1 01 24 AM" src="https://user-images.githubusercontent.com/7040122/147869737-d9b44545-11bd-4e61-836e-abbc24d6e3e4.png">

**Comment by crankywilson** [comment]
oh ok, thanks for the reply!  Do you work on Pylance?  I have a lot of comments...

1)  This worked great for me on my Linux Mint version of VSCode.  I'm getting the messages I want, and the feedback is very fast, even for slightly complexish scripts (> 1200 lines, containers of containers, etc.)
2)  This did not work on my Windows version of VSCode.  I'm guessing it's because my environment is old/corrupt.  I had like 10 different versions of Python installed all over my system and I tried to clean up.  Also, I probably have older stuff that hasn't been updated.  I'm betting if I uninstalled and re-installed the latest, I would have a better experience.  And 'not works' means I get some messages from pylance but not the above like in your screenshot.
3)  I used these tools a year or two ago, and I feel like I've seen a big improvement.  2 years ago, it seemed like the bigger the script got, the slower the analysis became to the point where it would be unusable at my script size (for like code completion suggestions, etc.).  Also, the complex types seemed to confuse the analysis before, but it seems to be rock solid now.  I felt like before Pyright was better for what I wanted but now I'm pretty happy with how my Linux (newer) environment is behaving at least with Pylance.
4)  I don't have any real world dev contacts - everything I pick up is from online and self-teaching can lead you to some, um 'noncanonical' habbits/conclusions...  Google leads me to believe that using the term linting as a synonym for code formatting is a misnomer.  ¯\_(ツ)_/¯ In any event, static type checking is much more to-the-point description.

**Reply by erictraut** [reply]
Yes, I'm the primary contributor to pyright. Pyright is the open-source type checker that Pylance is built upon. Pylance adds some additional convenience features like semantic highlighting, code refactoring, etc. on top of the core type checking.

I'm glad to hear that you've seen a big improvement over the past couple of years. The Python language server for VS Code was previously not based on a full static type checker. About a year ago, we switched from that older language server to Pylance, which now leverages the full power of pyright.
